### ISSUE TITLE
Minor test failure as underlying H5py library changed an exception from KeyError to TypeError

### ISSUE THREAD

**Issue by bnlawrence** [h5netcdf/h5netcdf#236]
#### What happened:

The test suite raised two errors

#### What you expected to happen:

No errors

#### Context:

Python 3.11.8
H5py 3.11.0
h5netcdf version 999 ??  (actually latest main)

#### Available Fix

There is an [available fix](https://github.com/bnlawrence/h5netcdf/commit/4d5477edb972aaa9f3b0931d6ed3f9576a7b80f7) in the main branch of my fork. I haven't done a pull request, because I don't know when and how the behaviour changed in the underlying library and/or python.

**Comment by kmuehlbauer** [comment]
Thanks @bnlawrence for raising this and sorry for the delay. This isn't raised in the current testsuite. Is this still an issue for you or can you give more details on your failing environment?

**Comment by kmuehlbauer** [comment]
we are testing:

```
      matrix:
        python-version: ["3.10", "3.12"]
        h5py-version: ["3.10", "3.11"]
        numpy-version: ["1"]
```
```
      matrix:
        python-version: ["3.11", "3.12"]
        h5py-version: ["3.11"]
        numpy-version: ["2"]
```

**Comment by kmuehlbauer** [comment]
@drew-parsons Please have a look at: https://github.com/h5netcdf/h5netcdf/actions/runs/11743663241/job/32717024125?pr=242

There is a fresh run with the current matrix. No issue there.

**Comment by drew-parsons** [comment]
If you're using the recent versions of h5py (>=3.8) then it's a strange discrepancy.

We can see h5py converting a KeyError into a TypeError at
https://github.com/h5py/h5py/blob/5c4871e37a08d7b030dc6c18df21f45cd714e6c6/h5py/_hl/group.py#L286
from https://github.com/h5py/h5py/pull/2092

So we could expect it to be a problem with recent h5py if h5netcdf still looks for KeyError at 
https://github.com/h5netcdf/h5netcdf/blob/b2ac2d74d85fadab786ee3760abe109964337dca/h5netcdf/tests/test_h5netcdf.py#L2245

**Comment by kmuehlbauer** [comment]
Thanks @drew-parsons. I'm having a hunch that these testcontext construction might not work as expected.

I'm afk, but will check this over the weekend. If you find the time you could remove the testcontext-stuff and call the latest.raises directly.

**Comment by drew-parsons** [comment]
The error I'm seeing is
```
_______________________________________________________________________________________________________________________ test_user_type_errors_new_api[testfile.nc] _______________________________________________________________________________________________________________________

tmp_local_or_remote_netcdf = '/tmp/pytest-of-me/pytest-13/test_user_type_errors_new_api_0/testfile.nc'

    def test_user_type_errors_new_api(tmp_local_or_remote_netcdf):
        enum_dict1 = dict(one=1, two=2, three=3, missing=254)
        enum_dict2 = dict(one=0, two=2, three=3, missing=255)
        with h5netcdf.File("test.nc", "w") as ds0:
            enum_type_ext = ds0.create_enumtype(np.uint8, "enum_t", enum_dict1)
            with h5netcdf.File(tmp_local_or_remote_netcdf, "w") as ds:
                ds.dimensions = {"enum_dim": 4}
                g = ds.create_group("subgroup")
                enum_type = ds.create_enumtype(np.uint8, "enum_t", enum_dict1)
    
                if tmp_local_or_remote_netcdf.startswith(remote_h5):
                    testcontext = pytest.raises(RuntimeError, match="Conflict")
                else:
                    testcontext = pytest.raises(KeyError, match="name already exists")
                with testcontext:
>                   ds.create_enumtype(np.uint8, "enum_t", enum_dict2)

h5netcdf/tests/test_h5netcdf.py:2247: 
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ 
h5netcdf/core.py:1390: in create_enumtype
    self._h5group[datatype_name] = et
/usr/lib/python3/dist-packages/h5py/_debian_h5py_serial/_hl/group.py:480: in __setitem__
    htype.commit(self.id, name, lcpl=lcpl)
h5py/_debian_h5py_serial/_objects.pyx:54: in h5py._debian_h5py_serial._objects.with_phil.wrapper
    ???
h5py/_debian_h5py_serial/_objects.pyx:55: in h5py._debian_h5py_serial._objects.with_phil.wrapper
    ???
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ 

>   ???
E   TypeError: Unable to commit datatype (name already exists)

h5py/_debian_h5py_serial/h5t.pyx:454: TypeError
______________________________________________________________________________________________________________________ test_user_type_errors_legacyapi[testfile.nc] ______________________________________________________________________________________________________________________

tmp_local_or_remote_netcdf = '/tmp/pytest-of-me/pytest-13/test_user_type_errors_legacyap0/testfile.nc'

    def test_user_type_errors_legacyapi(tmp_local_or_remote_netcdf):
        enum_dict1 = dict(one=1, two=2, three=3, missing=254)
        enum_dict2 = dict(one=0, two=2, three=3, missing=255)
        with legacyapi.Dataset("test.nc", "w") as ds0:
            enum_type_ext = ds0.createEnumType(np.uint8, "enum_t", enum_dict1)
            with legacyapi.Dataset(tmp_local_or_remote_netcdf, "w") as ds:
                ds.createDimension("enum_dim", 4)
                g = ds.createGroup("subgroup")
                enum_type = ds.createEnumType(np.uint8, "enum_t", enum_dict1)
                if tmp_local_or_remote_netcdf.startswith(remote_h5):
                    testcontext = pytest.raises(RuntimeError, match="Conflict")
                else:
                    testcontext = pytest.raises(KeyError, match="name already exists")
                with testcontext:
>                   ds.createEnumType(np.uint8, "enum_t", enum_dict1)

h5netcdf/tests/test_h5netcdf.py:2295: 
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ 
h5netcdf/core.py:1390: in create_enumtype
    self._h5group[datatype_name] = et
/usr/lib/python3/dist-packages/h5py/_debian_h5py_serial/_hl/group.py:480: in __setitem__
    htype.commit(self.id, name, lcpl=lcpl)
h5py/_debian_h5py_serial/_objects.pyx:54: in h5py._debian_h5py_serial._objects.with_phil.wrapper
    ???
h5py/_debian_h5py_serial/_objects.pyx:55: in h5py._debian_h5py_serial._objects.with_phil.wrapper
    ???
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ 

>   ???
E   TypeError: Unable to commit datatype (name already exists)

h5py/_debian_h5py_serial/h5t.pyx:454: TypeError
```

**Comment by drew-parsons** [comment]
The test passes if I ask it to expect `TypeError`, as in bnlawrence's patch

**Comment by kmuehlbauer** [comment]
@drew-parsons Thanks for sharing the stack trace. Here is the stacktrace I'm getting:

```python
tmp_local_or_remote_netcdf = '/tmp/pytest-of-kai/pytest-13/test_user_type_errors_new_api_0/testfile.nc'

    def test_user_type_errors_new_api(tmp_local_or_remote_netcdf):
        enum_dict1 = dict(one=1, two=2, three=3, missing=254)
        enum_dict2 = dict(one=0, two=2, three=3, missing=255)
        with h5netcdf.File("test.nc", "w") as ds0:
            enum_type_ext = ds0.create_enumtype(np.uint8, "enum_t", enum_dict1)
            with h5netcdf.File(tmp_local_or_remote_netcdf, "w") as ds:
                ds.dimensions = {"enum_dim": 4}
                g = ds.create_group("subgroup")
                enum_type = ds.create_enumtype(np.uint8, "enum_t", enum_dict1)
    
                if tmp_local_or_remote_netcdf.startswith(remote_h5):
                    with pytest.raises(RuntimeError, match="Conflict"):
                        ds.create_enumtype(np.uint8, "enum_t", enum_dict2)
                #else:
                #    with pytest.raises(KeyError, match="name already exists"):
>               ds.create_enumtype(np.uint8, "enum_t", enum_dict2)

test_h5netcdf.py:2247: 
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ 
../core.py:1390: in create_enumtype
    self._h5group[datatype_name] = et
/home/kai/mambaforge/envs/xr_312_np2/lib/python3.12/site-packages/h5py/_hl/group.py:480: in __setitem__
    htype.commit(self.id, name, lcpl=lcpl)
h5py/_objects.pyx:54: in h5py._objects.with_phil.wrapper
    ???
h5py/_objects.pyx:55: in h5py._objects.with_phil.wrapper
    ???
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ 

>   ???
E   KeyError: 'Unable to commit datatype synchronously (name already exists)'

h5py/h5t.pyx:454: KeyError

```

There's one difference in the output -> `synchronously`. This is directly coming from HDF5 issued in [H5TCommit2](https://github.com/HDFGroup/hdf5/blob/faec68f93608975527d9bb8b38d3aeda57644546/src/H5Tcommit.c#L160-L170). So I'm assuming your system uses a different HDF5 version which doesn't use the current H5TCommit2. 

The change in the error message came in by this PR https://github.com/HDFGroup/hdf5/pull/166 and is most likely present from HDF5 1.12. I also have no idea where the Python error is generated (somewhere in h5py I think). 

So this is definitely no h5netcdf issue, OTOH it would be good to have the test suite run for any supported HDF5 version.

**Comment by drew-parsons** [comment]
True, our HDF5 is 1.10.10.  It sounds plausible that accounts for the difference in behaviour.

**Comment by kmuehlbauer** [comment]
Closed by #244. Thanks for sticking to this until now, @drew-parsons.
